src/dev/flang/ast/Impl.java

```diff
@@ -522,13 +525,42 @@ AbstractType inferredType(Resolution res, AbstractFeature f)
     return switch (_kind)
       {
       case FieldDef    -> _initialValue.typeIfKnown();
-      case RoutineDef  -> _code.typeIfKnown();
+      case RoutineDef  -> typeFromRedefinedFeatures(f);
       case FieldActual -> typeFromInitialValues(res, f, false);
       default -> throw new Error("missing case "+_kind);
       };
   }
 
 
+  AbstractType typeFromRedefinedFeatures(AbstractFeature f)
+  {
+    AbstractType result;
+
+    if (PRECONDITIONS) require
+      (_kind == Kind.RoutineDef);
+
+    var rf = f.redefines()
+              .stream()
+              .filter(x -> x.state().atLeast(State.RESOLVED_TYPES))
```

Comment:
Instead of filtering what happens to have been resolved already (which depends on whether we are lucky or not), maybe explicitly resolve `x` via a call to `res.resolveTypes(x)`.